Jack/Seed.fs

```diff
@@ -32,87 +58,76 @@ module Seed =
     let random () : Seed =
-        Random().Next() |> ofInt32
+        let s = System.DateTimeOffset.UtcNow.Ticks + 2L * goldenGamma
+        { Value = mix64 s
+          Gamma = mixGamma s + goldenGamma }
 
-    /// The possible range of values returned from 'next'.
-    let range : int * int =
-        1, 2147483562
-    
     /// Returns the next pseudo-random number in the sequence, and a new seed.
-    let next (Seed (s1, s2)) : int * Seed =
-        // TODO remove crashUnless
-        crashUnless (s1 >= 0) "s1 >= 0"
-        let k    = s1 / 53668
-        let s1'  = 40014 * (s1 - k * 53668) - k * 12211
-        let s1'' = if s1' < 0 then s1' + 2147483563 else s1'
-    
-        crashUnless (s2 >= 0) "s2 >= 0"
-        let k'   = s2 / 52774
-        let s2'  = 40692 * (s2 - k' * 52774) - k' * 3791
-        let s2'' = if s2' < 0 then s2' + 2147483399 else s2'
-    
-        let z    = s1'' - s2''
-        let z'   = if z < 1 then z + 2147483562 else z
-    
-        z', Seed (s1'', s2'')
-    
+    let next (s : Seed) : int64 * Seed =
+        mix64 s.Value, nextSeed s
+
     /// Generate a random bigint in the specified range.
     let rec nextBigInt (lo : bigint) (hi : bigint) (seed : Seed) : bigint * Seed =
         if lo > hi then
             nextBigInt hi lo seed
         else
-            //
-            // Probabilities of the most likely and least likely result will differ
-            // at most by a factor of (1 +- 1/q). Assuming Seed is uniform, of
-            // course.
-            //
-            // On average, log q / log b more random values will be generated than
-            // the minimum.
-            //
-    
-            let genlo, genhi = range
-            let b = bigint genhi - bigint genlo + 1I
-    
-            let q = 1000I
-            let k = hi - lo + 1I
-            let magtgt = k * q
-    
-            // Generate random values until we exceed the target magnitude.
-            let rec loop mag v0 seed0 =
-                if mag >= magtgt then
-                    v0, seed0
-                else
-                    let x, seed1 = next seed0
-                    let v1 = v0 * b + (bigint x - bigint genlo)
-                    loop (mag * b) v1 seed1
-    
-            let v, seedN = loop 1I 0I seed
-
-            // TODO remove crashUnless
-            crashUnless (v >= 0I) "v >= 0I"
-            crashUnless (k >= 0I) "k >= 0I"
-            lo + v % k, seedN
+            let rec loop hilo (v0, seed0) =
+                let v0 = bigint (v0 : int64)
+                // hilo is the size of the exclusive range: 'hi - lo + 1'.
+                if hilo > bigint.Zero then
+                    // The range is a positive number – return a
+                    // number in range using modular arithmetic:
+                    // - http://codereview.stackexchange.com/a/29110/10541
+                    // - http://stackoverflow.com/a/1202706/467754
+                    lo + v0 % hilo, seed0
```

Comment:
So the issue here that I was worried about (and having the cast above is good as it makes it even clearer) is that `v0` will only ever be in the range `-2^63` to `2^63-1` because it's just a 64-bit number. So if `hilo` is greater than `2^64` then we'll never be able to generate numbers which cover the entire range of values.